### PR TITLE
feat: add new vlc-dag demos of poc

### PR DESCRIPTION
**Pull request by ai-chen2050**
A vlc-dag application.

The vlc-dag is base on the simple accumulator application. 
Please reference on crates/accumulator. But add new features as follow:
  * defined the vertex and edge of the event & clock propagation dag.
  * support the LMDB to persistence storage for now，
  * maybe time-series db,like influxDB, or related postgre more suitable. (TD)
  * Support increment state sync with peers using p2p communication protocol.
  * so we can use the vertex and edge from db to reconstruct clock propagation dag.

### CODE REVIEW

File: crates/vlc/src/lib.rs [hunk 1]
```diff
@@ -73,6 +79,52 @@ impl Clock {
             }
         }
     }
+
+    /// diff is bigger clock minus small clock
```
Comment (bufrr) [inline crates/vlc/src/lib.rs]: How do you know which one is smaller when you call this function?

File: crates/vlc/src/lib.rs [hunk 2]
```diff
@@ -73,6 +79,52 @@ impl Clock {
...
+        let mut ret = Clock::new();
+        for (id, v1) in &self.values {
+            let v2 = small.values.get(id).unwrap_or(&0);
+            ret.values.insert(*id, v1-v2);
+        }
+        ret
+    }
+
+    /// return index key of clock
+    pub fn index_key(&self) -> String {
```
Comment (bufrr) [inline crates/vlc/src/lib.rs:98]:
```suggestion
    pub fn index_key(&self) -> String {
    let mut key = String::new();
    for (index, value) in &self.values {
        key = format!("{}{}-{}-", key, index, value);
    }
    key
}
```

File: crates/vlc/src/lib.rs [hunk 3]
```diff
@@ -73,6 +79,52 @@ impl Clock {
...
+                ret.values.insert(*id, *v1);
+            } else {
+                ret.values.insert(*id, *v2);
+            }
+        }
+        ret
+    }
+
+    /// return true when all value is zero in clock dimensions
+    pub fn is_genesis(&self) -> bool {
```
Comment (bufrr) [inline crates/vlc/src/lib.rs:121]:
```suggestion
    pub fn is_genesis(&self) -> bool {
    let sum: u128 = self.values.values().sum();
    sum == 0
}
```

File: demos/vlc-dag/src/db_client/lldb_client.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+pub struct VLCLLDb {
+    env: Environment,
+    merge_log: DbHandle,
+    clock_infos: DbHandle,
+    cur_count: DbHandle,
+}
+
+impl VLCLLDb {
+    pub fn new(path: &str) -> Self {
+        const USER_DIR: u32 = 0o777;
```
Comment (bufrr) [inline demos/vlc-dag/src/db_client/lldb_client.rs]: We can make mode value as a parameter

File: demos/vlc-dag/src/db_client/lldb_client.rs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+        VLCLLDb { env, merge_log, clock_infos, cur_count }
+    }
+
+    pub(crate) fn add_clock_infos(&mut self, key: String, clock_info: ClockInfo) {
+        let txn = self.env.new_transaction().unwrap();
+        let db = txn.bind(&self.clock_infos);
+
+        let _ = db.set(&key, &serde_json::to_string(&clock_info).unwrap());
+        println!("【insert clock to DB】: \nkey: {},\nvalue: {:#?}", key, clock_info);
```
Comment (bufrr) [inline demos/vlc-dag/src/db_client/lldb_client.rs]: Do not use non-ASCII characters in logs

File: demos/vlc-dag/src/db_client/lldb_client.rs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        let db = txn.bind(&self.merge_log);
+
+        let _ = db.set(&key, &serde_json::to_string(&merge_log).unwrap());
+        println!("【insert merge_log to DB】:  \nkey: {},\nvalue: {:#?}", key, merge_log);
+    }
+
+    pub fn get_clock_info(&mut self, key: String) -> String {
+        let txn = self.env.new_transaction().unwrap();
+        let db = txn.bind(&self.clock_infos);
+        db.get::<&str>(&key).unwrap().to_string()
```
Comment (bufrr) [inline demos/vlc-dag/src/db_client/lldb_client.rs]: It may panic here if value does not exist

File: demos/vlc-dag/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,720 @@
...
+    clock_info: ClockInfo,
+    id: u128,
+    items: Vec<String>, // BtreeSet or HashSet can't keep event inserted order.
+    clock_to_eventid: HashMap<String, String>, // clock_id to event_id
+    message_id_set: BTreeSet<String>,
+}
+
+impl ServerState {
+    /// Create a new server state.
+    fn new(id: u128) -> Self {
```
Comment (bufrr) [inline demos/vlc-dag/src/lib.rs:154]: It's better to implement a `Default()` function

File: demos/vlc-dag/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,720 @@
...
+            },
+            id,
+            items: Vec::new(),
+            clock_to_eventid: HashMap::new(),
+            message_id_set: BTreeSet::new(),
+        }
+    }
+
+    /// Add items into the state. Returns true if resulting in a new state.
+    fn add(&mut self, items: Vec<String>) -> bool {
```
Comment (bufrr) [inline demos/vlc-dag/src/lib.rs:172]:
```suggestion
   fn add(&mut self, items: Vec<String>) -> bool {
    if items.is_empty() {
        return false;
    }

    for item in items.iter() {
        // filter replicate message id
        if !self.message_id_set.contains(item) {
            self.items.push(item.clone());
            self.message_id_set.insert(item.clone());
        }
    }

    self.clock_info.clock.inc(self.id);
    self.clock_info.count += 1;
    if let Some(last) = items.last() {
        self.clock_info.message_id = last.clone();
    }

    true
}
```

File: demos/vlc-dag/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,720 @@
...
+            }
+            None => {
+                self.clock_info.clock.merge(&vec![&other.clock_info.clock]);
+                let add = self.add(other.items.clone());
+                (add, add)
+            }
+        }
+    }
+
+    fn handle_event_trigger(&mut self, msg: EventTrigger) -> Option<ServerMessage> {
```
Comment (bufrr) [inline demos/vlc-dag/src/lib.rs:194]:
```suggestion
    fn handle_event_trigger(&mut self, msg: EventTrigger) -> Option<ServerMessage> {
    match self.clock_info.clock.partial_cmp(&msg.clock_info.clock) {
        Some(cmp::Ordering::Less) => {
            let self_clock_info = ClockInfo::new(self.clock_info.clock.clone(), self.id, self.clock_info.message_id.clone(), 0);
            let req = ServerMessage::DiffReq(DiffReq { from_clock: self_clock_info, to:  msg.clock_info.id });
            Some(req)
        }
        None => {
            let base = self.clock_info.clock.base_common(&msg.clock_info.clock);
            if base.is_genesis() || self.clock_to_eventid.get(&base.index_key()).map_or(false, |start_msg_id| get_suffix(&self.items, start_msg_id.to_string()).is_some()) {
                let req = ServerMessage::ActiveSync(ActiveSync { diffs: self.items.clone(), latest: self.clock_info.clone(), to: msg.clock_info.id });
                Some(req)
            } else {
                None
            }
        }
        _ => None,
    }
}
```

**Review by bufrr — CHANGES_REQUESTED** [review]

File: demos/vlc-dag/src/lib.rs (hunk 7 above)
Comment (ai-chen2050) [reply demos/vlc-dag/src/lib.rs:154]: Node id is a must parameter to construct a ServerState.

File: demos/vlc-dag/src/lib.rs (hunk 9 above)
Comment (ai-chen2050) [reply demos/vlc-dag/src/lib.rs:194]: Expose all of the conditions will be more information, explicitly.

**Review by bufrr — APPROVED** [review]

**Pull request merged**